File: dev_websockets/graphsettings.py
#!/usr/bin/python

"""
This file is part of the CINF Data Presentation Website
Copyright (C) 2012 Robert Jensen, Thomas Andersen and Kenneth Nielsen

The CINF Data Presentation Website is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License as published by the Free Software
Foundation, either version 3 of the License, or
(at your option) any later version.

The CINF Data Presentation Website is distributed in the hope
that it will be useful, but WITHOUT ANY WARRANTY; without even
the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License
along with The CINF Data Presentation Website.  If not, see
<http://www.gnu.org/licenses/>.
"""

import xml.etree.ElementTree

class graphSettings(object):
    """ This class parses the graphsettings from the graphsettings.xml file and
    returns them as the internal settings variables.
    """
    def __init__(self, typed=None, params={}, filepath='graphsettings.xml'):
        """ Initialise:

        Parameters:
        typed(str)      The type of graph, can be either: pressure, temperature,
                        morning_pressure, iv, xps, iss, masstime or massspectrum
        params(dict)    The parameters to be inserted in the settings
        """

        self.params = params
        # Create settings dictionary and input type and maybe id
        self.settings = {'typed':typed}
        if params.has_key('id') and params['id'] != '':
            self.settings['id'] = str(params['id'])


        gs = xml.etree.ElementTree.ElementTree()
        gs.parse(filepath)
        
        # Update with global settings
        global_settings = gs.find('global_settings')
        self._update_settings_with_xml(global_settings)

        # Update with graph settings
        if typed is not None:
            graph = [e for e in gs.findall('graph') if e.attrib['type'] == typed][0]
            self._update_settings_with_xml(graph)

    def _update_settings_with_xml(self, xml):
        """ Update the settings dictionary with the dictionary returned from
        xml_tree_to_associative_array
        """
        self.settings.update(self._xml_tree_to_assiciative_dictionary(xml))

    def _xml_tree_to_assiciative_dictionary(self, xml):
        """ Recursively turn xml tree into a \"dictionary tree\" """
        ret = {}
        for element in list(xml):
            if len(list(element)) == 0:
                ret[element.tag] = self._substitute_variables(element.text)
            else:
                ret[element.tag] = self._xml_tree_to_assiciative_dictionary(element)
        return ret

    def _substitute_variables(self, text):
        """ Substitute the {var} type variables in certain graphsettings
        entries with values from param
        """
        for key in self.params.keys():
            if self.params[key] != '':
                text = text.replace('{' + key + '}', str(self.params[key]))
        return text
    
    def get_settings(self):
        """ Return the settings dictionary """
        return self.settings

File: dev_bugs/graphsettings.php
<?php
  /*
    Copyright (C) 2012 Robert Jensen, Thomas Andersen and Kenneth Nielsen
    
    The CINF Data Presentation Website is free software: you can
    redistribute it and/or modify it under the terms of the GNU
    General Public License as published by the Free Software
    Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    The CINF Data Presentation Website is distributed in the hope
    that it will be useful, but WITHOUT ANY WARRANTY; without even
    the implied warranty of MERCHANTABILITY or FITNESS FOR A
    PARTICULAR PURPOSE.  See the GNU General Public License for more
    details.
    
    You should have received a copy of the GNU General Public License
    along with The CINF Data Presentation Website.  If not, see
    <http://www.gnu.org/licenses/>.
  */

ini_set('display_errors', 1); 
ini_set('log_errors', 1); 
ini_set('error_log', dirname(__FILE__) . '/error_log.txt'); 
error_reporting(E_ALL);

function xml_tree_to_assiciative_arrays($xml){
  $ret = Array();
  foreach ($xml->children() as $child){
    #echo($child->getName() . ":" . (string)$child->count() . "<br>");
    if ($child->count() > 0){
      #echo("yes<br>");
      $ret[$child->getName()] = xml_tree_to_assiciative_arrays($child);
	}
    else{
      #echo((string) $child . "<br>");
      $ret[$child->getName()] = (string) $child;
	}
  }
  
  return $ret;
}

function plot_settings($type, $params="", $ignore_invalid_type=False, $gs_file='graphsettings.xml'){
  # Load the system global defaults
  $gs_global = fopen("../global_settings.xml", 'r');
  $gs_global = fread($gs_global, filesize("../global_settings.xml"));
  $gs_xml_global = new SimpleXMLElement($gs_global);

  # Put the system global defaults in settings
  $settings = xml_tree_to_assiciative_arrays($gs_xml_global);

  # Load user graphsettings
  $gs = fopen($gs_file, 'r');
  $gs = fread($gs, filesize($gs_file));
  $gs_xml = new SimpleXMLElement($gs);

  # Update the settings with the USER global settings
  $user_global_settings = xml_tree_to_assiciative_arrays($gs_xml->global_settings);
  $settings = array_replace($settings, $user_global_settings);
  
  # Update with the graph specific settings
  $type_found = False;
  foreach ($gs_xml->graph as $g) {
    if ($g['type'] == $type) {
      $specific_settings = xml_tree_to_assiciative_arrays($g);
      $settings = array_replace($settings, $specific_settings);
      $type_found = True;
    }
  }
  # Check if the graph type was valid
  if (!$type_found and !$ignore_invalid_type){
    return NULL;
  }
  
  
  if (gettype($params) == 'array'){
    foreach($settings as $s_key => $s_value){
      foreach($params as $p_key => $p_value){
	$settings[$s_key] = str_replace('{'.$p_key.'}', $p_value, $settings[$s_key]);
      }
    }
  }
  
  return $settings;
  }
?>
